Azure Image OCR

dkotter opened the pull request:
### Description of the Change

Add support for [Azure image OCR](https://westcentralus.dev.cognitive.microsoft.com/docs/services/computer-vision-v3-ga/operations/56f91f2e778daf14a499f20d).

Adds a new setting that is turned off by default but can be turned on to run OCR scanning on images. When an image is initially uploaded, we run a few checks on that image. First, we check if the image matches one of the supported types (JPEG, PNG, GIF, BMP). If so, we then check if we have a previous image scan and if so, we check if we have either the handwriting or text tags set, with a high confidence level (above 90%). If all those checks pass, we then run OCR scanning.

If we get a successful response back, we then parse the text out of that response, save that text to `wp_content` for that image and then save the full response to post meta.

We also add scan/recan functionality to the media modal and single media edit screens. If a scan is run from here, we bypass our checks, as we assume if someone is manually starting a scan, they don't care about the checks. ~~Also added some better error handling to the scan/rescan buttons, so if a scan fails, we don't continue to show the loading icon but instead remove that, keep the button disabled and change the button text to say error (ideally we would add even better error handling in a separate PR).~~ _This is now in a separate PR: #231_

_TODO: last piece is to add `describedby` text into the content when needed. Need to support the block editor and the classic editor_

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected. -->

### Benefits

Images with text can now have that text automatically read and then inserted into the content, with the proper `describedby` tags. This helps provide more context for images, especially images that are screenshots of text (like social media posts).

### Possible Drawbacks

With this feature turned on, will cause a slight slowdown (seems like roughly 5 extra seconds), in image processing.

### Verification Process

<!--
What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., commands you ran, text you typed, buttons you clicked) and describe the results you observed.
-->

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests passed.


### Applicable Issues

#111

Commit:
Initial work on OCR processing. Add a new function, hooked to the generate attatchment metadata hook, that will kick off OCR processing. Add a new REST endpoint that will also utilize this same function. Add an OCR class that does all the processing. Currently only process PNG images.

Commit:
Add in setting to turn OCR on/off. Add recan functionality to media modal and single media edit screens. Bypass mime check if recan button or option is used. Add filter around approved media types

Commit:
Add helper function to find image that matches both file size and dimensions. Add support for the four file types azure supports. Some code cleanup

Commit:
Don't show OCR rescan options if that setting is turned off. Comment out supported image types besides PNG for now

Commit:
Better error handling for failed requests. Add callback to OCR rescan button, to populate the image description after success.

Commit:
Make sure we return proper values from our generate functions, to avoid errors. Remove local only code

Commit:
Move OCR scanning into the same function that does other image scanning. Pass the previously run image scan into our OCR function and utilize that to determine if the image needs OCR run or not.

Commit:
Don't store full response unless we have actual text to save

Commit:
Update docblocks

jeffpaul commented:
Noting here that @dinhtungdu is going to work on scaffolding in the Gutenberg bits

Commit:
Don't remove the OCR scan button if OCR is turned off. Turning OCR off should only turn off the automatic scanning, not the manual scanning

Commit:
Revert the extra error handling added here, as that's now in a separate PR

dinhtungdu committed:
feat: auto insert description generated by ocr after image block

dinhtungdu committed:
fix: only enqueue editor script if ocr is enabled

Commit:
Merge branch 'develop' into feature/azure-image-ocr

Commit:
Update OCR endpoint to the newly release v3.1

dinhtungdu committed:
feat: OCR modal and OCR sidebar button

Commit:
Fix tests

Commit:
If we modify the content using DOMDocument, make sure we account for encodings properly, so we don't end up with weird characters. Keep track of if we need to modify the content or not, so if we don't, we can just return the original and not risk messing anything up. Minor code formatting

Commit:
Switch to a different approach to handle encoding issues, to match what WordPress itself does in other places. This will hopefully be more lenient across environments and encoding types

helen commented on src/js/editor-ocr.js:
```diff
@@ -0,0 +1,148 @@
+	}, 10 ) );
+
+	return isOpen && <Modal title={__( 'ClassifAI detected text in your image' )}>
+		<p>{__( 'Would you like you insert it as a paragraph under you image block?' )}</p>
```
```suggestion
		<p>{__( 'Would you like you insert it as a paragraph under this image block?' )}</p>
```

Also, do these strings need to have the `classifai` textdomain?

dinhtungdu replied:
My bad, I forgot the textdomain, added it. Thank you!

helen commented:
This is feeling pretty good to me for a first run. I think what we might want to do is add a field to `wp_prepare_attachment_for_js()` that's a bool for whether `classifai_computer_vision_ocr` is non-empty (like `classifaiHasOcr` or something) and only show the prompt in that case, not just if the description field is populated because that description field could be populated manually or from unrelated EXIF data.

helen commented on src/js/editor-ocr.js:
```diff
@@ -0,0 +1,148 @@
+					<PanelBody title={__( 'ClassifAI' )} initialOpen={true}>
+						<PanelRow>
+							<Button onClick={() => insertDescription( clientId, attributes.id )} isPrimary>
+								{__( 'Insert description' )}
```
```suggestion
								{__( 'Insert scanned text into content' )}
```

Maybe more like this? I don't think being stored in the description field or being known as descriptive text is how you think of it in the editorial flow.

dinhtungdu committed:
fix: typo

dinhtungdu committed:
fix: only show modal if image is scanned by OCR

dinhtungdu committed:
fix: only show sidebar button if image has ocr text

dinhtungdu commented:
@helen I addressed your comment by exposing `classifai_computer_vision_ocr` in REST API response.

dinhtungdu committed:
try: replace sidebar button by toolbar button

dinhtungdu commented:
@helen in https://github.com/10up/classifai/pull/228/commits/4682dddf09475563505993e5c8bc0b9dc6244107, I replaced the sidebar button with a toolbar button. The UI is better, but it's not as descriptive and explicit as the sidebar button is. Or should we use text instead of icon, for example `Insert scanned text`?

dinhtungdu committed:
fix: update toolbar icon

jeffpaul committed:
Merge branch 'develop' into feature/azure-image-ocr

dinhtungdu committed:
fix: ocr status in media api response

dinhtungdu committed:
Merge branch 'feature/azure-image-ocr' of github.com:10up/classifai into feature/azure-image-ocr

dinhtungdu commented:
@helen I fixed the REST API response issue. This PR is working on my live site.

helen committed:
Merge branch 'develop' into feature/azure-image-ocr

Commit:
Make sure the OCR manual scan button is added independently of the smart cropping button. Make sure we have HTML elements before trying to add event handlers to them

Commit:
Make sure we don't show the scan buttons on the single edit view, only the modal view

helen committed:
Insert a verse block instead of a paragraph.

This way any additional line breaks will keep the single block with its ID intact.

Also massages the message in the modal.

helen committed:
Go back to paragraph block for now.

Didn't realize verse was a pre without line wrapping :(

dinhtungdu committed:
fix: use group block for scanned text block

dinhtungdu committed:
ci: use composer v1

dinhtungdu committed:
fix: only allow one ocr block per image

helen committed:
Create block style for group for editor styling purposes

CSS isn't yet building into `dist`

helen committed:
Commas, sigh.

helen committed:
Better CSS for highlight border

helen committed:
Highlight related image/OCR block when editor is focused on the other.

This is in-progress, see code comments for needs

Commit:
Add the classnames utility in order to merge existing classnames with the classname we want to add. Minor linting fixes

Commit:
Remove the ocr-related-block class when elements aren't selected anymore

helen committed:
Add `classifai_ocr_text_post_args` filter

This allows for a number of customizations, such as being able to store the text results in a different field in case you use post_content extensively in the editorial workflow already, or set other meta, or update the alt text based on the text results, and so forth.

Also adds the full $scan data as context for the `classfai_ocr_text` fitler.

Commit:
Make sure we don't constantly set and remove the class if an image block is selected

Commit:
Merge branch 'feature/azure-image-ocr' of github.com:10up/classifai into feature/azure-image-ocr

Commit:
Filter post content before it's saved to remove our OCR class that is only used for editor highlighting

dinhtungdu committed:
fix: switch to use internal style

dinhtungdu committed:
fix: more sensible timeout

dinhtungdu committed:
fix: deal with different backgrounds

dinhtungdu committed:
Merge pull request #257 from 10up/try/ocr-internal-style

fix: switch to use internal style

helen reviewed (approved):
Nice work here, everybody!
